### PR TITLE
Improve the usage of Sized ByteBufferPool

### PR DESCRIPTION
**Pull request by gregw**
Followup work to https://github.com/jetty/jetty.project/pull/12020.

### CODE REVIEW

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1406]: I would change this constructor to also take `boolean direct, long maxSize`, this would make it obvious this class wants a size and directness.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: Is that `8192L` related to the default in `ByteBufferPool.Sized`? At least this should be made a constant with a meaningful name.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1455]: This check can go as `ByteBufferPool.Sized` does not allow 0 size.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: You've just added `ByteBufferPool.Sized.acquire(int size)` that could be used here.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: You don't need to call `asMutable()` anymore.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (lorban) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java:26]: `getByteBufferPool()` should be removed, but that could be done in another PR and the method be just deprecated for now.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
Comment (lorban) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:226]: It looks like this pattern could be extracted in a specific helper class that would take an init lambda in its constructor, maybe `LazyAtomicReference`?

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1455]: It does now.  We need 0 size so that can indicate that aggregation is not to be done, but we still might need a pool for some copies.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1406]: We have that constructor below.  This is for somebody that has a ByteBufferPool and wants a DynamicCapacity buffer, but is OK with all the defaults.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1455]: I think there is only one usage of this (see lines 2004-2006 in append and the javdoc of aggregationSize above).   Perhaps we don't need this mode, but we still want the ability to make a buffer that will only retain and never aggregate.

File: jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3Configuration.java
Comment (lorban) [inline jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3Configuration.java]: `ByteBuf bytes`? I think you meant `{@link IO#DEFAULT_BUFFER_SIZE} bytes`

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:226]: Nah, that's not a great idea in practice.

File: jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3Configuration.java
Comment (lorban) [reply jetty-core/jetty-http3/jetty-http3-common/src/main/java/org/eclipse/jetty/http3/HTTP3Configuration.java]: Fixed.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java:26]: Removed and replaced all its usages with get Input/Output equivalent.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:1455]: Since the sized pool now allows 0 byte buffers, this check can stay.

**Comment by gregw** [comment]
@lorban @sbordet Is this PR needed anymore? It is not needed for the HttpContent stuff, but  I think there are other good cleanups that it achieves.    It is not absolutely vital, but still worth some consideration?

**Comment by lorban** [comment]
@gregw I think most of the cleanups here are nice additions, except for the change to `Components`: I now believe we should not expose a couple of sized pools as they make us mistakenly conflate the input/output buffer sizes with unrelated buffer sizes.

I have also opened https://github.com/jetty/jetty.project/pull/12060 (which lives on top of https://github.com/jetty/jetty.project/pull/12020) that modifies the `HttpContent`-related APIs to use `ByteBufferPool.Sized` instead of the non-sized variant plus a few minor cleanups, fixes and realignments of ee10 and ee11.

**Comment by gregw** [comment]
I'm converting this back to a draft and will wait until you land your writeTo PRs and then will refresh.

**Comment by lorban** [comment]
Superseded by #12060

**Pull request closed without merging**
